doc/life-of-a-request.md

```diff
@@ -0,0 +1,164 @@
+option.
+
+It then determines whether there is policy associated with the remote peer's address
+and/or the original destination address. This policy may, for instance, pre-determine how
+the connection should be handled (i.e. with regard to TLS/HTTP/etc).
+
+### Protocol determination
+
+If protocol has not been pre-determined by policy, the proxy attempts to discover the
+protocol by reading data from the client. Note that this detection can break
+server-speaks-first protocols like MySQL and SMTP, necessitating the need to disable such
+detection by policy.
+
+In order to detect the connection's session protocol transparently (i.e. without policy),
+the proxy reads from the connection, buffering data until the proxy determines how to
+handle the connection.
+
+#### TLS
+
+If the connection begins with a TLS[[1]](#tls-q1) client handshake, data is read until
+until a full `ClientHello` message is buffered. The proxy determines whether it should
+terminate the TLS connection or treat the connection as an opaque TCP stream. The proxy
+may terminate TLS when a Conduit-specific extension[[2]](#tls-q2) exists in the client's
+hello, or it may terminate TLS if it is configured with a certificate for the
+client-requested name (i.e. for external/ingress).
+
+If the proxy elects to terminate TLS, the client's ALPN[[3]](#tls-q3) advertisement is
+used to determine which protocol should be used. If the ALPN advertisement does not
+include a proxy-supported protocol (i.e. `http` or `h2`), then the proxy handles the
+decrypted stream as opaque TCP[4](#tls-q4).
+
+##### Questions
+
+<a name="tls-q1"></a>
+1. What's the minimum version of the TLS protocol we support? I assume we don't handle SSL
+    3.0, etc. due to ClientHello requirement
+<a name="tls-q2"></a>
+2. How do TLS extensions play into this detection? are they conduit-specific or are we
+   reusing existing public extensions?
+<a name="tls-q3"></a>
+3. Can we reasonably expect clients to support ALPN? Do we have to support detecting
+   protocols for TLS requests that don't advertise a protocol?
+<a name="tls-q4"></a>
+4. It's not immediately clear to me how the proxy can "bridge" two TLS
+   connections--propagating ALPN information could be complex: we might want to issue a
+   `ClientHello` to the destination before issuing a ServerHello to the initiatior. This is
+   mostly germane for ingress.
+
+
+#### HTTP
+
+If the connection begins with an HTTP/1 request-line, the connection is handled as HTTP/1.
+
+If the connection begins with a HTTP/2 Prior-knowledge preface, the connection is handled
+as HTTP/2.
+
+
+## TCP Forwarding
+
+If, after protocol detection (and potentially TLS decryption), the connection's protocol
+is not supported by the proxy, the connection is simply forwarded to the original
+destination.
+
+Policy may require that the forwarded connections are TLS'd by the proxy (i.e. for mutual
+authentication).
```

Comment:
I don't know what this means but I think we can assume that either we're terminating TLS and at least one side of the connection is plaintext, or we're not terminating TLS and it doesn't matter. We can't terminate TLS and then "re-TLS" the connection so the application can do mutual auth because TLS is specifically designed to prevent that.

Reply:
I was imagining an ingress-like use case where we terminate some external TLS and then use internally-authorized connections between the ingress proxy and the destination. Probably not an important case to consider at the moment.

Reply:
I see. Re-reading what you wrote, I see that you just mean that we might accept an incoming connection that may or may not be over TLS, and then we might forward the connection, but we might need to add TLS to the connection when forwarding. I understand now.